### PR TITLE
Require C++17

### PR DESCRIPTION
**Pull request by masterleinad**
This pull request still leaves the checks
- DEAL_II_HAVE_CXX17_BESSEL_FUNCTIONS
- DEAL_II_HAVE_CXX17_LEGENDRE_FUNCTIONS
- DEAL_II_CXX14_CONSTEXPR_BUG

but assumes we can use all other C++17 features. There are also a couple of places with comments that C++17 could be used that this pull request doesn't touch.

### CODE REVIEW

**Review by drwells — APPROVED** [review]

**Comment by bangerth** [comment]
A valid question is what we can do with C++17 that we can't do right now. Principally, this really only covers `if constexpr` and structured bindings. Is that worth requiring new compilers? What did, for example, Ubuntu 18.04 ship with?

(We could also use `std::optional` and `std::variant`, but we have BOOST fall-backs for that. @Rombur and @masterleinad told us today that you can't use execution policies with current clang compilers. I am personally partial to variadic template fold expressions, but there are not that many places where this would be useful and we work around this successfully right now. Perhaps the syntactically nicest things is [initializers in if and switch statements](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0305r1.html), but we've done fine without them so far as well.)

I'm not opposed to this change, but I'm also not entirely convinced that we really gain very much.

**Comment by masterleinad** [comment]
https://godbolt.org/z/conKPcfzd shows that we can use execution policies for `clang 11` and higher which mean that we still need a feature test or compiler version check if we wanted to use those (`clang 10` and lower seem to have some problems pulling in TBB on godbolt).

**Comment by Rombur** [comment]
The issue is not really clang but libc++ (see [here](https://en.cppreference.com/w/cpp/compiler_support/17)) clang on linux works because it often uses libstdc++.

**Comment by masterleinad** [comment]
For me, the main advantage is being able to use `if constexpr` instead of SFINAE and compatibility with other dependencies that require C++17 now (`Kokkos`, `ArborX`, `CGAL`, `Trilinos`).

The default compiler for Ubuntu 18.04 is `GCC 7.4. 0` but standard support has ended on 31 May 2023 already. I think we can discuss minimum compiler requirements later.

**Comment by marcfehling** [comment]
`taskflow` requires C++17, see https://github.com/taskflow/taskflow#supported-compilers

**Comment by Rombur** [comment]
c++17 is not very exciting but I do like the consistency of bumping the require C++ standard every three years. So I am in favor of this PR.

**Review by tamiko — APPROVED** [review]

**Review by marcfehling — APPROVED** [review]

**Comment by marcfehling** [comment]
`tidy` complains about cgal:
```
/dealii/source/cgal/intersections.cc:86:5: error: use '= default' to define a trivial default constructor [modernize-use-equals-default,-warnings-as-errors]
/dealii/source/grid/grid_generator_cgal.cc:102:9: error: use 'using' instead of 'typedef' [modernize-use-using,-warnings-as-errors]
```

**Comment by bangerth** [comment]
I addressed the two issues pointed out by `tidy` in separate PRs.

**Comment by bangerth** [comment]
The vote this morning was 8 yes and 2 abstain. So we will merge this once all of the checks succeed.

File: include/deal.II/base/std_cxx17/algorithm.h [hunk 1]
```diff
@@ -17,6 +17,9 @@
 
 #include <deal.II/base/config.h>
 
+DEAL_II_WARNING(
+  "This file is deprecated. Simply use the corresponding C++14 header <algorithm>.")
```
Comment (bangerth) [inline include/deal.II/base/std_cxx17/algorithm.h:21]: I think this ought to be C++17. Same for `optional.h` below, as well as the other files in this directory.

Comment (masterleinad) [reply include/deal.II/base/std_cxx17/algorithm.h:21]: Can if fix this in a follow-up if the CI passes?

Comment (bangerth) [reply include/deal.II/base/std_cxx17/algorithm.h:21]: Yes.

**Comment by masterleinad** [comment]
There are a couple of `MatrixFree` tests in `Jenkins: serial` that are failing with "out of memory". I can't reproduce it locally, though, and it seems unrelated to the changes in this pull request.

**Pull request merged**